### ISSUE TITLE
Add support for "add to home screen" web apps

### ISSUE THREAD

**Issue by johanbove** [falzonv/discreet-launcher#16]
As a web app user I want to add my most used web apps to the "home screen" with my browser, so that I have immediate access to these apps. 

Currently Discreet Launcher only supports native apps.

No idea how much effort this is, but I am personally missing this. Otherwise great idea and useful app! Thanks.

**Comment by falzonv** [comment]
Hello Johan,

Thank you very much for your feedback, I am glad you like the app!

I am not very familiar with how web apps should looks like in a phone launcher.
Reading your message, my understanding is that it should looks like a normal application but would open the default web browser to a specific URL. It would be a sort of "web shortcut" app. Is it right?

This is indeed a very interesting idea, I will think about how to do it and keep you updated.
Not sure about the exact workload either, but since I do this on my spare time it will probably take some time.

Best regards.

**Comment by johanbove** [comment]
Thanks for the reply and the consideration. Good documentation on how "progressive web apps" work can be found here: https://docs.microsoft.com/en-us/microsoft-edge/progressive-web-apps-chromium/

And of course, take your time! If I knew Android coding and Java, I would have a look at this myself, but I'm only a JS coder.

**Comment by falzonv** [comment]
Hello,

Thank you for the documents, it was an interesting reading and, indeed, not really what I was thinking about.

In parallel, I made several tests and was able to make some progress (no code yet at this stage).
Both to keep you updated and for me as reference, I will write a sort of report here:

---

First, when testing on my phone, I noticed that Firefox was showing the option to add a web app when using my stock Android launcher but was disabling this option when using Discreet Launcher (switching the default launcher in the settings was switching the option accordingly). So the launcher must announce, in a way or another, that he is able to manage web apps.

Then, when clicking the option (with the stock Android launcher), Firefox calls directly the launcher and asks it to add the web app, nothing special is done by the system after Firefox has sent the request. The next question is how the launcher should receive this request? In which form is it sent?

Before trying to dig in Firefox source code to understand what it does, I found the application [WebApps](https://f-droid.org/en/packages/com.tobykurien.webapps) on F-Droid which is quite more focused on the topic. At the bottom of its [WebAppActivity](https://github.com/tobykurien/WebApps/blob/3541336deb05065959464ea405ec5e2af620c86e/app/src/main/java/com/tobykurien/webapps/activity/WebAppActivity.xtend) class, the interesting stuff is done by the method <code>addShortcut()</code> which calls the Android method <code>[ShortcutManagerCompat.requestPinShortcut()](https://developer.android.com/reference/androidx/core/content/pm/ShortcutManagerCompat#requestPinShortcut(android.content.Context,%20androidx.core.content.pm.ShortcutInfoCompat,%20android.content.IntentSender))</code>.

Looking at the [code of requestPinShortcut()](https://androidx.tech/artifacts/core/core/1.3.2-source/androidx/core/content/pm/ShortcutManagerCompat.java.html) is interesting but also a bit disappointing to be honest:
- Before Android Oreo (8.0) it is done with a broadcast named *com.android.launcher.action.INSTALL_SHORTCUT* so we should be good by listening for it (although I did not find much info about this broadcast in the Android documentation...).
- Starting with Android Oreo, it is quite more obscure: the method <code>[ShortcutManager.requestPinShortcut()](https://developer.android.com/reference/android/content/pm/ShortcutManager#requestPinShortcut(android.content.pm.ShortcutInfo,%20android.content.IntentSender))</code> is called (same name without "Compat") and [its code](https://android.googlesource.com/platform/frameworks/base/+/master/core/java/android/content/pm/ShortcutManager.java) is calling again another method <code>IShortcutService.requestPinShortcut()</code> but this is an abstract method (no code) and I have not found its implementation yet. So for now, no idea about what <code>requestPinShortcut()</code> is sending and how it could be received...

In the process, I discovered how it is checked if the launcher supports web apps.
There is a method <code>[isRequestPinShortcutSupported()](https://developer.android.com/reference/android/content/pm/ShortcutManager#isRequestPinShortcutSupported())</code> which seems to act the same way than <code>requestPinShortcut()</code> (that is, send *com.android.launcher.action.INSTALL_SHORTCUT* before Oreo and something unknown starting with Oreo) but apparently with empty data. So I guess once I have found what is supposed to be received, both issues should be solved at the same time.

To sum up, here is what needs to be done:
- Find a way to tell other applications that Discreet Launcher supports web apps (will be needed to test with Firefox anyway, as the option is disabled otherwise)
   - Might even be find two ways to support both before and after Oreo
- Find a way to receive the "requestPinShortcut()" request and data (web app name, icon and how to open it)
   - Might even be find two ways to support both before and after Oreo
- Add the web app to the list of applications (this should also be stored more statically, in a similar way than the favorites list)
- In v1.2.0 a "long click to open settings" feature was added, for a web app it should be replaced by a dialog allowing to remove the web app from the list (since there is no system "Uninstall" button as for native applications)

---

That is all for now, we will see where it goes :-)

Best regards.

**Comment by falzonv** [comment]
Hello Johan,

Status update!
Although it has not been released yet, **web apps are now supported with Android versions before Oreo (8.0)**.

Of course you probably have a more recent Android version, but that is still a very good news because most of the required code to support web apps is common to all versions and is now ready. The only missing bit is the way to receive the request which is not clear after Oreo (details in my long message above if you are interested), I am still working on it.

Meanwhile, **you can find below some screenshots** of the result with an Android version 5.0.

Some remarks:
- Testing with [Wiktionary](https://en.m.wiktionary.org), I discovered that web apps were again different than what I understood :-)
At the end this a good thing because what I implemented is a solution to manage web shortcuts, which happens to also work fine with web apps. So once everything is ready, both will be supported!
- Web apps (and shortcuts) are fully integrated in the applications list of Discreet Launcher. It means that they can be added to the favorites panel or to the notification menu (new in v2.1.0) as any other application.
- A long click on a web app displays an option to remove it (instead of the system settings for native applications).
- For now, a generic icon is used for web apps and shortcuts and it will probably stay like this for some time. This is because the custom icon can be retrieved only when receiving the request and must be stored as a bitmap file if we want to display it again later. Although it is far from being impossible, I am not very happy with this solution at the moment.

**Request to add the Wiktionary web app after clicking the Home(+) button**
![2021-03-27_web_app1](https://user-images.githubusercontent.com/4698531/112720037-686feb00-8efc-11eb-8746-20cbf5afa534.jpg)

**It is successfully added (see at the end of the list due to "W")**
![2021-03-27_web_app2](https://user-images.githubusercontent.com/4698531/112720116-c997be80-8efc-11eb-8165-db0cfbefcd30.jpg)

**The web app can also be added to the favorites panel**
![2021-03-27_web_app3](https://user-images.githubusercontent.com/4698531/112720158-f1872200-8efc-11eb-9aa9-731e409c3302.jpg)

**Or to the notification menu** (new in v2.1.0)
![2021-03-27_web_app4](https://user-images.githubusercontent.com/4698531/112720192-298e6500-8efd-11eb-8e8c-50fb64281c68.jpg)

**And of course, it works when clicking the icon!** (no Firefox interface)
![2021-03-27_web_app5](https://user-images.githubusercontent.com/4698531/112720246-6fe3c400-8efd-11eb-928d-196220ac9224.jpg)

Best regards.

**Comment by johanbove** [comment]
This is awesome @falzonv ! Merci beaucoup.
I'm on Android 10 - so for me, all should be alright. Looking forward to testing this.

**Comment by falzonv** [comment]
Avec plaisir, ce n'était pas facile mais très formateur :-)
*(suite en anglais pour la cohérence du ticket)*

**This is now working for all Android versions from 5 to 11!**
More precisely, it was successfully tested with Android 5 (Firefox), Android 8.1 (Firefox, Chrome) and Android 11 (Firefox).
Now I can only hope that it will work for you as well.

This second part was definitely not easy, previously I wrote:

> The **only missing bit** is the way to receive the request which is not clear after Oreo

Well, this "only missing bit" was in fact very big and it took me ~10 hours to make it work (understand how Android was doing the request, refactor most of what I had already done, find tricks to make the shortcut usable from the notification, etc.).
It would have been much simpler if they stayed with the previous implementation... anyway, now it works!

This will be part of **release v2.2.0** that I will try to finalize and **push to GitHub tomorrow evening**.
Once it is pushed on GitHub, the **F-Droid build and publishing process is automatic but usually takes 4 to 5 days to complete**.
So yeah, it will take some waiting time now...

Finally and to keep it as reference, I will wrote below another report for this second part.
It may help someone wanting to do the same in the future, because all this part about receiving shortcuts on the launcher side is surprisingly obscure in the Android documentation.

Best regards.

---

So, continuing from the first report...

Previously, I had found that starting with Android Oreo (8.0), the method <code>[ShortcutManager.requestPinShortcut()](https://developer.android.com/reference/android/content/pm/ShortcutManager#requestPinShortcut(android.content.pm.ShortcutInfo,%20android.content.IntentSender))</code> is used, and it is calling <code>IShortcutService.requestPinShortcut()</code> from the abstract class <code>[IOShortcutService](https://android.googlesource.com/platform/frameworks/base/+/master/core/java/android/content/pm/IShortcutService.aidl)</code>.

Yesterday I was able to find the source code of <code>[ShortcutService](https://android.googlesource.com/platform/frameworks/base/+/master/services/core/java/com/android/server/pm/ShortcutService.java)</code> which is the implementation of <code>IOShortcutService</code>.

Here, the method <code>requestPinShortcut()</code> calls a method <code>requestPinItem()</code> which perform all kinds of checks and finally call <code>requestPinItemLocked()</code> from the class <code>[ShortcutRequestPinProcessor](https://cs.android.com/android/platform/superproject/+/master:frameworks/base/services/core/java/com/android/server/pm/ShortcutRequestPinProcessor.java)</code>.

In this method <code>requestPinItemLocked()</code>, two things are performed:
1. Check if the launcher supports shortcuts with the method <code>getRequestPinConfirmationActivity()</code>
2. If yes, prepare the shortcut that will be send to the launcher with the method <code>requestPinShortcutLocked()</code>

Searching for <code>getRequestPinConfirmationActivity()</code> in the same file shows that it is also used for the method <code>[isRequestPinShortcutSupported()](https://developer.android.com/reference/android/content/pm/ShortcutManager#isRequestPinShortcutSupported())</code> that I had already noticed Thursday, interesting.
Looking at the code of <code>getRequestPinConfirmationActivity()</code>, it is using the method <code>injectGetPinConfirmationActivity()</code> (which send us back to the <code>[ShortcutService](https://cs.android.com/android/platform/superproject/+/master:frameworks/base/services/core/java/com/android/server/pm/ShortcutService.java)</code> class) to analye the package of the default launcher (according to system settings), looking for the action *android.content.pm.action.CONFIRM_PIN_SHORTCUT* in the manifest.

I added this action to the manifest and now the "Add shortcut" option of Firefox is not grayed anymore with Discreet Launcher.
=> Point 1 above is solved!

Regarding point 2, the method <code>requestPinShortcutLocked()</code> starts by doing several checks and prepares the <code>[ShortcutInfo](https://developer.android.com/reference/android/content/pm/ShortcutInfo)</code> object that will be sent to the launcher.
This object is cloned from one that is kept internally by Android but the <code>Intent</code> telling how to start the shortcut is removed from the clone, this information is not given to the launcher (which probably explains "Locked" in the name).
Once the request is ready, it is sent back to the method <code>requestPinItemLocked()</code> which should now forward it to the launcher.

This is done using again another method from <code>ShortcutRequestPinProcessor</code>, called <code>startRequestConfirmActivity()</code> which finally send an <code>Intent</code> containing the prepared request with the action *android.content.pm.action.CONFIRM_PIN_SHORTCUT* (the one previously declared in the manifest).

In order to receive this, I had to create a new activity (file <code>ActivityShortcut.java</code> in the source code) dedicated for this purpose.
Adding a basic text message to display when this activity starts, I tried to add a web app and the message got displayed!

The next step was to retrieve the <code>ShortcutInfo</code> sent by <code>startRequestConfirmationActivity()</code>, and extract from it the name of the shortcut and enough information to be able to launch it later (with legacy shortcut, it is simply an intent).
Easy for the name using <code>shortcut.getShortLabel()</code>, quite more complex for the information to launch because I wanted something that could be written to a file (to save the shortcut if the user restarts Discreet Launcher, or the smartphone).

Starting with Android Oreo (8) and according to the <code>[ShortcutInfo.getIntent()](https://developer.android.com/reference/android/content/pm/ShortcutInfo#getIntent())</code> documentation:

> Launcher apps **cannot** see the intent. If a ShortcutInfo is obtained via LauncherApps, then this method will always return null. Launchers can only start a shortcut intent with LauncherApps#startShortcut.

Looking at <code>[LauncherApps.startShortcut()](https://developer.android.com/reference/android/content/pm/LauncherApps#startShortcut(java.lang.String,%20java.lang.String,%20android.graphics.Rect,%20android.os.Bundle,%20android.os.UserHandle))</code>, three information are really needed (the others can be null):
- The name of the package which requested the shortcut, obtained with <code>shortcut.getPackage()</code>
- The ID of the shortcut, obtained with <code>shortcut.getId()</code>
- And a <code>[UserHandle](https://developer.android.com/reference/android/os/UserHandle)</code> object, obtained with <code>shortcut.getUserHandle()</code>

For the two first ones, they are <code>String</code> objects so no issue to store them in a file.

Regarding the third one I want to store it also as a <code>String</code> (concatenated with the other and using separators).
Hopefully, the method <code>UserHandle.toString()</code> gives a textual representation of the format <code>UserHandle{0}</code> where 0 seems the most common ID (represents the system) but it seems that it can be a different value in some cases.
Knowing this, I simply parsed the result of <code>toString()</code> (remove what is not useful and try to get the Integer) and applied 0 as default if this was not possible for some reasons.

With all these information and many other code refactoring, it was finally working from both the drawer list and the favorites panel.
But not from the notification menu...

This is because a proper <code>Intent</code> is really needed for the buttons of the notification, it cannot use directly the method <code>LauncherApps.startShortcut()</code>.
Since we do not have the shortcut intent that Android keeps secret, I used my new <code>ActivityShortcut</code> class as a sort of "transfer area": an intent to call this class is provided to the notification and in this intent we add the 3 required parameters (package, id, user). When the button is clicked, the <code>ActivityShortcut</code> class receives the intent with the parameters and finally call <code>LauncherApps.startShortcut()</code> for a successful launch.

I don't think many people will read so far, if you did thank you for your time and I hope it was interesting!

**Comment by falzonv** [comment]
Hello again,

Finally the last things to finalize were quicker than I thought and release v2.2.0 implementing this feature is now pushed to GitHub.

Please note that my previous comment is still true:
> Once it is pushed on GitHub, the F-Droid build and publishing process is automatic but **usually takes 4 to 5 days to complete**.

Best regards.

**Comment by falzonv** [comment]
Hello,

I am confused because this is taking a very long time on F-Droid side.
Apparently v2.2.0 has finally been built yesterday so it should be available soon but still not on the website for now...

Maybe it is my reference which is wrong, I am always assuming 4 to 5 days but in practice I observed more 4 to 8 days (well, even 9 days in our case).
So the publishing delay after push would be 4 to 8 days for F-Droid and 1 to 2 days for Google Play (but the comparison is not really fair because Google has a simpler work to do and many more servers to do it).

Best regards.

**Comment by johanbove** [comment]
I received v2.2.0 yesterday on my phone.

Tested the creation of the "home screen icon for web apps" - and it worked with Brave browser (uses Chromium). But the action did "crash" Discreet Launcher and I got the standard Android launcher until it restarted Discreet Launcher. But then the web app icon was there. Only thing that didn't seem to stick is the web app icon. The default Discreet Launcher is used in the application list.

**Comment by falzonv** [comment]
Hello,

Thank you for the test and feedback!

Quite surprised about the crash itself because all possible things seemed tested (well, obviously not...), but maybe even more surprised that, even with the crash, the web app addition still worked.
If this was fine up to this point, then the crash probably happened during the application list update. However, the exact same update function is called when starting the launcher and there was no issue the second time.

Some questions to better understand the issue:
- Did this issue happened only the first time you added a web app or does it still happens?
- If still happening, does it happens with all web apps or only with some specifics?

Regarding the icon, this is what I was describing when saying the following for Android < 8 :
> For now, a generic icon is used for web apps and shortcuts and it will probably stay like this for some time. This is because the custom icon can be retrieved only when receiving the request and must be stored as a bitmap file if we want to display it again later. Although it is far from being impossible, I am not very happy with this solution at the moment.

In the meantime I have added support with Android >= 8 and this is managed in a totally different way (quite more complicated).

Now that these parts are working (if we put aside the crash remaining to solve), I can indeed think about how to add the real icons and use the generic only as a fallback.
Seeing that another user already created a ticket for this, the "icons for web apps" part will be followed in #21.

Best regards.

**Comment by johanbove** [comment]
The "crash"  was only on the first addition of a web app. Consequtive web app additions are not crashing and I see that it detects the new app and it reinits the app list. Nice job!
I can live without the actual icons until a later release. And I think we can close this issue. Thank you!
Also: thanks for the mention in the release notes, but I don't deserve any credit really :-)

**Comment by falzonv** [comment]
Cool, I am glad that it is working fine! :-)
The issue on first addition is still strange but I was not able to reproduce it from a clean install so we will indeed close the issue.

For me you deserve it for the idea: I was not even aware about web apps before, so no chance to see them in Discreet Launcher if you had not opened this ticket in the first place ;-)

Have a nice evening!

**Comment by falzonv** [comment]
Final update on this ticket!

While working on ticket #21 (icons for web apps), I was able to solve the strange issue you faced at the beginning so this is corrected in the next release v2.5.0.

To explain briefly, the way to return from "ActivityShortcut" (called when receiving a creation request) to the home screen was working in most cases but in some cases it failed and quit the application instead.
I was not able to find exactly in which cases so I completely modified the way to return to the home screen instead and now there should be no issue.

Best regards.
